GitExtUtils/GitUI/Theming/TabControlPaintContext.cs

```diff
@@ -0,0 +1,309 @@
+        private readonly Rectangle[] _tabRects;
+        private readonly string[] _tabTexts;
+        private readonly Color[] _tabBackColors;
+        private readonly Size _size;
+        private readonly bool _failed;
+
+        private static int ImagePadding { get; } = DpiUtil.Scale(6);
+        private const int SelectedTabPadding = 2;
+        private const int BorderWidth = 1;
+
+        public TabControlPaintContext(TabControl tabs, PaintEventArgs e)
+        {
+            _mouseCursor = tabs.PointToClient(Cursor.Position);
+            _graphics = e.Graphics;
+            _clipRectangle = e.ClipRectangle;
+            _size = tabs.Size;
+            _parentBackColor = GetParentBackColor(tabs);
+            _selectedIndex = tabs.SelectedIndex;
+            _tabCount = tabs.TabCount;
+            _font = tabs.Font;
+            _imageSize = tabs.ImageList?.ImageSize ?? Size.Empty;
+            _enabled = tabs.Enabled;
+
+            try
+            {
+                _tabTexts = Enumerable.Range(0, _tabCount)
+                    .Select(i => tabs.TabPages[i].Text)
+                    .ToArray();
+                _tabImages = Enumerable.Range(0, _tabCount)
+                    .Select(i => GetTabImage(tabs, i))
+                    .ToArray();
+                _tabRects = Enumerable.Range(0, _tabCount)
+                    .Select(tabs.GetTabRect)
+                    .ToArray();
+                _tabBackColors = Enumerable.Range(0, _tabCount)
+                    .Select(i => tabs.TabPages[i].BackColor)
+                    .ToArray();
+            }
+            catch (ArgumentOutOfRangeException)
+            {
+                // Workaround probable bug in .NET framework, known example:
+                // tabCtrl.GetTabRect[tabCtrl.SelectedIndex] may throw ArgumentOutOfRangeException
+                // https://github.com/gitextensions/gitextensions/pull/7213#issuecomment-554760531
+                _failed = true;
+
+                // Set these to null explicitly to satisfy nullability checking. We will always verify
+                // _failed before dereferencing these.
+                _tabTexts = null!;
+                _tabImages = null!;
+                _tabRects = null!;
+                _tabBackColors = null!;
+            }
+        }
+
+        public void Paint()
+        {
+            if (_failed)
+            {
+                return;
+            }
+
+            using var canvasBrush = new SolidBrush(_parentBackColor);
+            _graphics.FillRectangle(canvasBrush, _clipRectangle);
+
+            RenderSelectedPageBackground();
+
+            IEnumerable<int> pageIndices;
+            if (_selectedIndex.IsWithin(0, _tabCount))
+            {
+                // render tabs in pyramid order with selected on top
+                pageIndices = Enumerable.Range(0, _selectedIndex)
+                    .Concat(Enumerable.Range(_selectedIndex, _tabCount - _selectedIndex).Reverse());
```

Comment:
Linq is known to have perf-related overhead. Paint operations we'd want to be as lean and mean as possible.
Perhaps worth considering switching this to a plain `for` or `foreach` loop.

Reply:
I cannot imagine loop -> linq overhead per 1 iteration to be to so huge to become comparable with painting ~ 20 x 100 px rectangle of background + icon + text.